src/meta_memcache/base/base_cache_pool.py

```diff
@@ -0,0 +1,314 @@
+import base64
+from abc import ABC, abstractmethod
+from collections import defaultdict
+from typing import (
+    Any,
+    Callable,
+    DefaultDict,
+    Dict,
+    Final,
+    List,
+    Optional,
+    Set,
+    Tuple,
+    Union,
+)
+
+from meta_memcache.base.base_serializer import BaseSerializer
+from meta_memcache.base.base_write_failure_tracker import BaseWriteFailureTracker
+from meta_memcache.base.connection_pool import ConnectionPool
+from meta_memcache.base.memcache_socket import MemcacheSocket
+from meta_memcache.errors import MemcacheError, MemcacheServerError
+from meta_memcache.protocol import (
+    ENDL,
+    Conflict,
+    Flag,
+    IntFlag,
+    Key,
+    MetaCommand,
+    Miss,
+    NotStored,
+    Success,
+    TokenFlag,
+    Value,
+)
+from meta_memcache.settings import MAX_KEY_SIZE
+
+
+class BaseCachePool(ABC):
+    def __init__(
+        self,
+        serializer: BaseSerializer,
+        binary_key_encoding_fn: Callable[[Key], bytes],
+        write_failure_tracker: Optional[BaseWriteFailureTracker] = None,
+    ) -> None:
+        self._serializer = serializer
+        self._binary_key_encoding_fn = binary_key_encoding_fn
+        self._write_failure_tracker: Final = write_failure_tracker
+
+    @abstractmethod
+    def _get_pool(self, key: Key) -> ConnectionPool:
+        ...
+
+    def _encode_key(self, key: Key) -> Tuple[bytes, bool]:
+        if key.is_unicode or len(key.key) > MAX_KEY_SIZE:
+            return base64.b64encode(self._binary_key_encoding_fn(key)), True
+        else:
+            return key.key.encode("ascii"), False
+
+    def _build_cmd(
+        self,
+        command: MetaCommand,
+        key: Key,
+        size: Optional[int] = None,
+        flags: Optional[Set[Flag]] = None,
+        int_flags: Optional[Dict[IntFlag, int]] = None,
+        token_flags: Optional[Dict[TokenFlag, bytes]] = None,
+    ) -> bytes:
+        encoded_key, is_binary = self._encode_key(key)
+        cmd = [command.value, encoded_key]
+        if size is not None:
+            cmd.append(str(size).encode("ascii"))
+        cmd_flags = []
+        if is_binary:
+            cmd_flags.append(Flag.BINARY.value)
+        if flags:
+            cmd_flags.extend(sorted(flag.value for flag in flags))
+        if int_flags:
+            for flag, int_value in int_flags.items():
+                cmd_flags.append(flag.value + str(int_value).encode("ascii"))
+        if token_flags:
+            for flag, bytes_value in token_flags.items():
+                cmd_flags.append(flag.value + bytes_value)
+        cmd.extend(sorted(cmd_flags))
```

Comment:
Again, this is just so command flag order is deterministic. I might just seed the set ordering in python in the tests... This definitely needs to be fixed to avoid hurting performance for testability.

Reply:
Agree, my question is because sorting is a bit expensive and if it's for testability we have an improvement point here, nothing serious though